File: routes/checklists.py
"""Onboarding checklist route handlers."""
import json as _json

from routes import add_log
import db as _db
import checklist_csv as _csv


class ChecklistsMixin:
    """Mixin for onboarding checklist endpoints."""

    # ── Template ──────────────────────────────────────────────────

    def _checklist_template(self):
        """GET /api/checklists/template — parsed CSV template."""
        try:
            tmpl = _csv.load_checklist_template()
        except FileNotFoundError:
            self._json_response({"error": "checklist CSV not found"}, 503)
            return
        self._json_response(tmpl)

    # ── List / Create ─────────────────────────────────────────────

    def _list_checklists(self, qs=None):
        """GET /api/checklists[?status=active|completed|archived]"""
        status = (qs or {}).get("status", [None])[0] if qs else None
        rows = _db.list_checklists(status)

        # Attach progress % and linked site name to each row
        try:
            tmpl = _csv.load_checklist_template()
            all_steps = tmpl["steps"]
        except Exception:
            all_steps = []

        # Preload all linked site names in one shot
        cached = {str(s["id"]): s.get("name", "") for s in _db.get_cached_sites()}

        result = []
        for row in rows:
            cid = row["id"]
            data = _db.get_checklist_data(cid)
            applicable = _checklist_applicable_steps(row, all_steps)
            total_app = len(applicable)
            completed = sum(
                1 for s in applicable
                if data.get(s["id"], {}).get("value") not in (None, "", "0", "false", False)
            )
            pct = round(completed / total_app * 100) if total_app else 0

            linked_name = None
            if row.get("mainwp_site_id"):
                linked_name = cached.get(str(row["mainwp_site_id"]), f"site #{row['mainwp_site_id']}")

            result.append({
                **row,
                "progress_pct": pct,
                "progress_completed": completed,
                "progress_total": total_app,
                "linked_site_name": linked_name,
            })

        self._json_response(result)

    def _create_checklist(self, body):
        """POST /api/checklists"""
        client_name = (body.get("client_name") or "").strip()
        if not client_name:
            self._json_response({"error": "client_name required"}, 400)
            return
        site_url = (body.get("site_url") or "").strip()
        is_new_client = bool(body.get("is_new_client", True))
        is_redesign = bool(body.get("is_redesign", False))
        is_new_site = bool(body.get("is_new_site", True))
        custom_host = (body.get("custom_host") or "").strip()

        cid = _db.create_checklist(client_name, site_url, is_new_client,
                                   is_redesign, is_new_site, custom_host)
        add_log("Checklists", "ok", f"Checklist created: {client_name} (id={cid})")
        self._json_response({"ok": True, "id": cid})

    # ── Get / Update / Delete ─────────────────────────────────────

    def _get_checklist(self, checklist_id):
        """GET /api/checklists/<id>"""
        row = _db.get_checklist(int(checklist_id))
        if not row:
            self._json_response({"error": "not found"}, 404)
            return
        data = _db.get_checklist_data(int(checklist_id))
        self._json_response({**row, "data": data})

    def _update_checklist(self, checklist_id, body):
        """PATCH /api/checklists/<id>"""
        row = _db.get_checklist(int(checklist_id))
        if not row:
            self._json_response({"error": "not found"}, 404)
            return
        _db.update_checklist(int(checklist_id), **body)
        self._json_response({"ok": True})

    def _delete_checklist(self, checklist_id):
        """DELETE /api/checklists/<id>"""
        row = _db.get_checklist(int(checklist_id))
        if not row:
            self._json_response({"error": "not found"}, 404)
            return
        _db.delete_checklist(int(checklist_id))
        add_log("Checklists", "ok", f"Checklist deleted: id={checklist_id}")
        self._json_response({"ok": True})

    # ── Cell ──────────────────────────────────────────────────────

    def _save_checklist_cell(self, checklist_id, body):
        """POST /api/checklists/<id>/cell"""
        step_id = str(body.get("step_id", "")).strip()
        if not step_id:
            self._json_response({"error": "step_id required"}, 400)
            return
        value = str(body.get("value", ""))
        _db.save_checklist_cell(int(checklist_id), step_id, value)
        self._json_response({"ok": True})

    # ── Link / Unlink ─────────────────────────────────────────────

    def _link_checklist(self, checklist_id, body):
        """POST /api/checklists/<id>/link"""
        try:
            mainwp_site_id = int(body.get("mainwp_site_id"))
        except (TypeError, ValueError):
            self._json_response({"error": "mainwp_site_id required"}, 400)
            return
        _db.link_checklist_to_mainwp(int(checklist_id), mainwp_site_id)
        add_log("Checklists", "ok",
                f"Checklist {checklist_id} linked to site {mainwp_site_id}")
        self._json_response({"ok": True})

    def _unlink_checklist(self, checklist_id):
        """POST /api/checklists/<id>/unlink"""
        _db.unlink_checklist(int(checklist_id))
        self._json_response({"ok": True})

    # ── Available unlinked sites ──────────────────────────────────

    def _checklist_available_sites(self):
        """GET /api/checklists/available-sites — sites not already linked to a checklist."""
        cached = _db.get_cached_sites()
        # Collect all currently linked mainwp_site_ids
        linked_ids = {
            row["mainwp_site_id"]
            for row in _db.list_checklists()
            if row.get("mainwp_site_id")
        }
        sites = [
            {"id": s["id"], "name": s.get("name", ""), "url": s.get("url", "")}
            for s in cached
            if s["id"] not in linked_ids
        ]
        self._json_response(sites)


# ── Helpers ────────────────────────────────────────────────────────────────


def _checklist_applicable_steps(checklist_row: dict, all_steps: list) -> list:
    """Return steps that are 'applicable' (not requiring a condition the checklist doesn't have)."""
    is_new_client = bool(checklist_row.get("is_new_client", True))
    is_redesign = bool(checklist_row.get("is_redesign", False))
    is_new_site = bool(checklist_row.get("is_new_site", True))
    custom_host = (checklist_row.get("custom_host") or "").strip().lower()
    is_custom_host = bool(custom_host and custom_host not in ("", "siteground"))

    result = []
    for step in all_steps:
        if step.get("requires_new_client") and not is_new_client:
            continue
        if step.get("requires_current_client") and is_new_client:
            continue
        if step.get("requires_redesign") and not is_redesign:
            continue
        if step.get("requires_new_site") and not is_new_site:
            continue
        if step.get("requires_custom_host") and not is_custom_host:
            continue
        result.append(step)
    return result

File: static/js/checklists.js
// ============================================================================
// Onboarding Checklists — client onboarding step-by-step tracker
// ============================================================================
// Extracted from the original monolithic index.html <script> during the
// ES-module split.

import { modalPush, modalRemove, _esc } from './dom.js';
import { apiCall } from './api.js';
import { registerRoute } from './router.js';

import { dbg } from './debug.js';
dbg('module', 'checklists.js loaded');

        // ── Onboarding Checklists event wiring ───────────────────────────────────

        // Back button
        document.getElementById('obcBackBtn').addEventListener('click', () => {
            _obcCurrentId = null;
            document.getElementById('obcDetail').style.display = 'none';
            document.getElementById('obcList').style.display = '';
            loadChecklists();
        });

        // Edit toggle
        document.getElementById('obcEditCheck').addEventListener('change', e => {
            document.getElementById('obcDetailContainer').classList.toggle('obc-locked', !e.target.checked);
        });

        // Filter chips
        document.querySelectorAll('.obc-chip').forEach(chip => {
            chip.addEventListener('click', () => {
                document.querySelectorAll('.obc-chip').forEach(c => c.classList.remove('active'));
                chip.classList.add('active');
                _obcFilter = chip.dataset.status;
                loadChecklists();
            });
        });

        // New Onboarding button → show inline modal
        document.getElementById('obcNewBtn').addEventListener('click', () => {
            _showObcNewModal();
        });

        function _showObcNewModal() {
            // Create a lightweight inline modal
            const overlay = document.createElement('div');
            overlay.className = 'modal-overlay visible';
            overlay.style.zIndex = '310';
            overlay.innerHTML = `
                <div class="modal" style="max-width:440px;width:95vw;">
                    <h2>New Onboarding Checklist</h2>
                    <div class="form-group">
                        <label>Client Name *</label>
                        <input type="text" class="field-input" id="obcNewName" placeholder="Acme Corp" autofocus>
                    </div>
                    <div class="form-group">
                        <label>Site URL</label>
                        <input type="url" class="field-input" id="obcNewUrl" placeholder="https://example.com">
                    </div>
                    <div class="form-group">
                        <label>Client Type</label>
                        <div style="display:flex;gap:16px;margin-top:4px;">
                            <label style="display:flex;align-items:center;gap:6px;font-size:13px;cursor:pointer;">
                                <input type="radio" name="obcClientType" value="new" checked> New Client
                            </label>
                            <label style="display:flex;align-items:center;gap:6px;font-size:13px;cursor:pointer;">
                                <input type="radio" name="obcClientType" value="current"> Current Client
                            </label>
                        </div>
                    </div>
                    <div class="form-group">
                        <label>Project Type</label>
                        <div style="display:flex;gap:16px;margin-top:4px;">
                            <label style="display:flex;align-items:center;gap:6px;font-size:13px;cursor:pointer;">
                                <input type="radio" name="obcProjectType" value="new_site" checked> New Site/Domain
                            </label>
                            <label style="display:flex;align-items:center;gap:6px;font-size:13px;cursor:pointer;">
                                <input type="radio" name="obcProjectType" value="redesign"> Redesign
                            </label>
                        </div>
                    </div>
                    <div class="form-group">
                        <label>Host</label>
                        <div style="display:flex;gap:16px;align-items:center;margin-top:4px;flex-wrap:wrap;">
                            <label style="display:flex;align-items:center;gap:6px;font-size:13px;cursor:pointer;">
                                <input type="radio" name="obcHost" value="siteground" checked> Siteground
                            </label>
                            <label style="display:flex;align-items:center;gap:6px;font-size:13px;cursor:pointer;">
                                <input type="radio" name="obcHost" value="other"> Other:
                            </label>
                            <input type="text" id="obcNewCustomHost" class="field-input" style="width:140px;display:none;" placeholder="host name">
                        </div>
                    </div>
                    <div class="modal-buttons" style="margin-top:20px;">
                        <button class="btn" id="obcNewCancelBtn">Cancel</button>
                        <button class="btn btn-primary" id="obcNewCreateBtn">Create</button>
                    </div>
                </div>`;
            document.body.appendChild(overlay);

            // Show/hide custom host input
            overlay.querySelectorAll('input[name="obcHost"]').forEach(r => {
                r.addEventListener('change', () => {
                    document.getElementById('obcNewCustomHost').style.display =
                        overlay.querySelector('input[name="obcHost"]:checked').value === 'other' ? '' : 'none';
                });
            });

            const close = () => { overlay.remove(); modalRemove('obcNewModal'); };
            modalPush('obcNewModal', close);

            overlay.addEventListener('click', e => { if (e.target === overlay) close(); });
            document.getElementById('obcNewCancelBtn').addEventListener('click', close);
            document.getElementById('obcNewCreateBtn').addEventListener('click', async () => {
                const name = document.getElementById('obcNewName').value.trim();
                if (!name) { document.getElementById('obcNewName').focus(); return; }
                const hostRadio = overlay.querySelector('input[name="obcHost"]:checked').value;
                const customHost = hostRadio === 'other' ? (document.getElementById('obcNewCustomHost').value.trim()) : 'siteground';
                const projectType = overlay.querySelector('input[name="obcProjectType"]:checked').value;
                const clientType = overlay.querySelector('input[name="obcClientType"]:checked').value;
                close();
                await _obcCreateNew({
                    client_name: name,
                    site_url: document.getElementById('obcNewUrl').value.trim(),
                    is_new_client: clientType === 'new',
                    is_redesign: projectType === 'redesign',
                    is_new_site: projectType === 'new_site',
                    custom_host: customHost,
                });
            });

            // Focus
            setTimeout(() => document.getElementById('obcNewName').focus(), 50);
        }

        // ── Onboarding Checklists ─────────────────────────────────────────────────

        let _obcTemplate = null;  // { sections, steps }
        let _obcFilter = 'active';
        let _obcCurrentId = null;
        let _obcCurrentData = {};  // { step_id: { value, completed_at } }

        async function loadChecklists() {
            const container = document.getElementById('obcListContainer');
            container.innerHTML = '<div style="padding:16px;color:var(--text-muted);font-size:13px;">Loading…</div>';

            // Ensure we have the template (sections/steps from CSV)
            if (!_obcTemplate) {
                try { _obcTemplate = await apiCall('/checklists/template'); } catch(e) { /* ignore */ }
            }

            try {
                const qs = _obcFilter ? `?status=${_obcFilter}` : '';
                const rows = await apiCall(`/checklists${qs}`);
                renderObcList(rows, container);
            } catch(e) {
                container.innerHTML = `<div class="empty-state"><h3>Error</h3><p>${_esc(e.message)}</p></div>`;
            }
        }

        function renderObcList(rows, container) {
            if (!rows.length) {
                const label = _obcFilter || 'onboarding checklists';
                container.innerHTML = `<div class="empty-state"><h3>No ${_esc(label)} checklists</h3><p>Click "+ New Onboarding" to create one.</p></div>`;
                return;
            }
            let html = `<table class="obc-list-table"><thead><tr>
                <th>Client</th><th>Site URL</th><th>Progress</th><th>Status</th><th>Linked Site</th><th>Updated</th><th></th>
            </tr></thead><tbody>`;
            rows.forEach(row => {
                const pct = row.progress_pct ?? 0;
                const badge = `<span class="obc-badge obc-badge-${row.status}">${_esc(row.status)}</span>`;
                const linked = row.linked_site_name ? _esc(row.linked_site_name) : '<span style="color:var(--text-muted);font-size:12px;">— not linked —</span>';
                const updated = (row.updated_at || '').slice(0, 16).replace('T', ' ');
                html += `<tr class="obc-list-row" data-obc-id="${row.id}">
                    <td style="font-weight:500;">${_esc(row.client_name)}</td>
                    <td style="font-size:12px;color:var(--text-muted);">${_esc(row.site_url || '')}</td>
                    <td>
                        <div class="obc-progress-wrap">
                            <div class="obc-progress-bar"><div class="obc-progress-fill" style="width:${pct}%"></div></div>
                            <span class="obc-progress-label">${row.progress_completed}/${row.progress_total}</span>
                        </div>
                    </td>
                    <td>${badge}</td>
                    <td>${linked}</td>
                    <td style="font-size:12px;color:var(--text-muted);">${updated}</td>
                    <td>
                        <button class="btn" style="font-size:11px;padding:3px 10px;" data-obc-open="${row.id}">Open</button>
                        <button class="btn" style="font-size:11px;padding:3px 10px;" data-obc-delete="${row.id}">Delete</button>
                    </td>
                </tr>`;
            });
            html += '</tbody></table>';
            container.innerHTML = html;

            // Row click → open; button overrides
            container.querySelectorAll('tr.obc-list-row').forEach(tr => {
                tr.addEventListener('click', e => {
                    if (e.target.closest('button')) return;
                    openObcDetail(parseInt(tr.dataset.obcId));
                });
            });
            container.querySelectorAll('[data-obc-open]').forEach(btn => {
                btn.addEventListener('click', e => { e.stopPropagation(); openObcDetail(parseInt(btn.dataset.obcOpen)); });
            });
            container.querySelectorAll('[data-obc-delete]').forEach(btn => {
                btn.addEventListener('click', e => {
                    e.stopPropagation();
                    const id = parseInt(btn.dataset.obcDelete);
                    if (!confirm('Delete this checklist and all its progress?')) return;
                    apiCall(`/checklists/${id}`, 'DELETE').then(() => loadChecklists());
                });
            });
        }

        async function openObcDetail(checklistId) {
            _obcCurrentId = checklistId;
            document.getElementById('obcList').style.display = 'none';
            document.getElementById('obcDetail').style.display = '';

            // Reset edit toggle
            document.getElementById('obcEditCheck').checked = false;
            document.getElementById('obcDetailContainer').classList.add('obc-locked');

            document.getElementById('obcDetailContainer').innerHTML =
                '<div style="padding:16px;color:var(--text-muted);font-size:13px;">Loading…</div>';

            try {
                const [cl, availSites] = await Promise.all([
                    apiCall(`/checklists/${checklistId}`),
                    apiCall('/checklists/available-sites'),
                ]);
                _obcCurrentData = cl.data || {};
                renderObcDetail(cl, availSites);
            } catch(e) {
                document.getElementById('obcDetailContainer').innerHTML =
                    `<div class="empty-state"><h3>Error</h3><p>${_esc(e.message)}</p></div>`;
            }
        }

        function renderObcDetail(cl, availSites) {
            const tmpl = _obcTemplate || { sections: [], steps: [] };
            const allSteps = tmpl.steps || [];

            // Condition flags
            const isNewClient = !!cl.is_new_client;
            const isRedesign = !!cl.is_redesign;
            const isNewSite = !!cl.is_new_site;
            const customHost = (cl.custom_host || '').trim().toLowerCase();
            const isCustomHost = !!(customHost && customHost !== 'siteground');

            // Build condition badges
            const condBadges = [
                isNewClient ? 'New Client' : 'Current Client',
                isRedesign ? 'Redesign' : 'New Site',
                isCustomHost ? `Host: ${_esc(cl.custom_host)}` : 'Siteground',
            ].map(t => `<span class="obc-cond-badge">${t}</span>`).join('');

            // Link/unlink control
            let linkHtml;
            if (cl.mainwp_site_id) {
                const linkedName = availSites.__linked_name || `site #${cl.mainwp_site_id}`;
                linkHtml = `
                    <div class="obc-link-row">
                        <span class="obc-link-label">Linked to MainWP:</span>
                        <strong style="font-size:13px;">${_esc(String(cl.mainwp_site_id))}</strong>
                        <button class="btn" id="obcUnlinkBtn" style="font-size:11px;">Unlink</button>
                    </div>`;
            } else {
                const options = availSites.map(s =>
                    `<option value="${s.id}">${_esc(s.name)} — ${_esc(s.url || '')}</option>`
                ).join('');
                linkHtml = `
                    <div class="obc-link-row">
                        <span class="obc-link-label">Link to MainWP site:</span>
                        <select class="field-input obc-link-select" id="obcLinkSelect">
                            <option value="">— select site —</option>
                            ${options}
                        </select>
                        <button class="btn btn-primary" id="obcLinkBtn" style="font-size:11px;">Link</button>
                    </div>`;
            }

            // Progress calculation
            const applicableSteps = allSteps.filter(s => _obcStepApplies(s, cl));
            const completedCount = applicableSteps.filter(s => {
                const v = (_obcCurrentData[s.id] || {}).value;
                return v && v !== '0' && v !== 'false';
            }).length;
            const totalCount = applicableSteps.length;
            const pct = totalCount ? Math.round(completedCount / totalCount * 100) : 0;

            // Header
            let html = `
            <div class="obc-detail-header">
                <div class="obc-detail-title">${_esc(cl.client_name)}</div>
                ${cl.site_url ? `<div class="obc-detail-url">${_esc(cl.site_url)}</div>` : ''}
                <div class="obc-cond-badges">${condBadges}</div>
                ${linkHtml}
                <div class="obc-detail-progress">
                    <div class="obc-detail-progress-bar">
                        <div class="obc-detail-progress-fill" id="obcProgressFill" style="width:${pct}%"></div>
                    </div>
                    <div class="obc-detail-progress-label" id="obcProgressLabel">${completedCount} of ${totalCount} applicable steps complete</div>
                </div>
            </div>`;

            // Steps by section
            const sections = tmpl.sections || [];
            sections.forEach(section => {
                const sectionSteps = allSteps.filter(s => s.section === section);
                if (!sectionSteps.length) return;
                html += `<div class="obc-section">
                    <div class="obc-section-title">${_esc(section)}</div>`;
                sectionSteps.forEach(step => {
                    html += _obcRenderStep(step, cl);
                });
                html += '</div>';
            });
            // Steps with no section
            const noSection = allSteps.filter(s => !s.section);
            if (noSection.length) {
                html += `<div class="obc-section">`;
                noSection.forEach(step => { html += _obcRenderStep(step, cl); });
                html += '</div>';
            }

            const container = document.getElementById('obcDetailContainer');
            container.innerHTML = html;

            // Wire up link/unlink buttons
            const linkBtn = document.getElementById('obcLinkBtn');
            if (linkBtn) {
                linkBtn.addEventListener('click', async () => {
                    const sel = document.getElementById('obcLinkSelect');
                    if (!sel.value) return;
                    await apiCall(`/checklists/${_obcCurrentId}/link`, 'POST', { mainwp_site_id: parseInt(sel.value) });
                    openObcDetail(_obcCurrentId);
                });
            }
            const unlinkBtn = document.getElementById('obcUnlinkBtn');
            if (unlinkBtn) {
                unlinkBtn.addEventListener('click', async () => {
                    await apiCall(`/checklists/${_obcCurrentId}/unlink`, 'POST', {});
                    openObcDetail(_obcCurrentId);
                });
            }

            // Wire up step inputs
            container.querySelectorAll('[data-step-id]').forEach(el => {
                const stepId = el.dataset.stepId;
                if (el.type === 'checkbox') {
                    el.addEventListener('change', () => _obcSaveCell(stepId, el.checked ? '1' : ''));
                } else if (el.tagName === 'SELECT') {
                    el.addEventListener('change', () => _obcSaveCell(stepId, el.value));
                } else {
                    el.addEventListener('blur', () => _obcSaveCell(stepId, el.value));
                }
            });

            // Wire up description toggles
            container.querySelectorAll('[data-desc-toggle]').forEach(btn => {
                btn.addEventListener('click', () => {
                    const desc = document.getElementById(`obcDesc-${btn.dataset.descToggle}`);
                    if (desc) desc.classList.toggle('visible');
                });
            });
        }

        function _obcStepApplies(step, cl) {
            const isNewClient = !!cl.is_new_client;
            const isRedesign = !!cl.is_redesign;
            const isNewSite = !!cl.is_new_site;
            const customHost = (cl.custom_host || '').trim().toLowerCase();
            const isCustomHost = !!(customHost && customHost !== 'siteground');
            if (step.requires_new_client && !isNewClient) return false;
            if (step.requires_current_client && isNewClient) return false;
            if (step.requires_redesign && !isRedesign) return false;
            if (step.requires_new_site && !isNewSite) return false;
            if (step.requires_custom_host && !isCustomHost) return false;
            return true;
        }

        function _obcRenderStep(step, cl) {
            const applies = _obcStepApplies(step, cl);
            const disabledClass = applies ? '' : ' obc-step-disabled';
            const saved = (_obcCurrentData[step.id] || {}).value || '';

            let inner = '';
            if (step.field_type === 'bool') {
                const checked = (saved === '1' || saved === 'true') ? 'checked' : '';
                inner = `<div class="obc-step-check">
                    <input type="checkbox" data-step-id="${_esc(step.id)}" ${checked}>
                    <span class="obc-step-name">${_esc(step.name)}</span>
                </div>`;
            } else if (step.field_type === 'select') {
                const options = (step.options || []).map(o =>
                    `<option value="${_esc(o)}" ${saved === o ? 'selected' : ''}>${_esc(o)}</option>`
                ).join('');
                inner = `<div class="obc-step-select">
                    <span class="obc-step-select-label">${_esc(step.name)}</span>
                    <select class="field-input obc-step-dropdown" data-step-id="${_esc(step.id)}">
                        <option value="">—</option>${options}
                    </select>
                </div>`;
            } else {
                // text or url
                inner = `<div class="obc-step-text">
                    <span class="obc-step-text-label">${_esc(step.name)}</span>
                    <input type="${step.field_type === 'url' ? 'url' : 'text'}"
                        class="obc-step-input"
                        data-step-id="${_esc(step.id)}"
                        value="${_esc(saved)}"
                        placeholder="…">
                </div>`;
            }

            const descBtn = step.description
                ? `<button class="obc-step-desc-btn" data-desc-toggle="${_esc(step.id)}" title="Show instructions">?</button>`
                : '';
            const descEl = step.description
                ? `<div class="obc-step-desc" id="obcDesc-${_esc(step.id)}">${_esc(step.description)}</div>`
                : '';

            return `<div class="obc-step${disabledClass}">
                <div style="flex:1;min-width:0;">${inner}${descEl}</div>
                ${descBtn}
            </div>`;
        }

        async function _obcSaveCell(stepId, value) {
            if (!_obcCurrentId) return;
            _obcCurrentData[stepId] = { value, completed_at: value ? new Date().toISOString() : null };
            try {
                await apiCall(`/checklists/${_obcCurrentId}/cell`, 'POST', { step_id: stepId, value });
                _obcUpdateProgress();
            } catch(e) { /* silent */ }
        }

        function _obcUpdateProgress() {
            const tmpl = _obcTemplate || { steps: [] };
            // We need the current checklist conditions — read from DOM badges
            // Simpler: re-fetch from stored _obcCurrentData + template
            const fill = document.getElementById('obcProgressFill');
            const label = document.getElementById('obcProgressLabel');
            if (!fill || !label) return;
            const applicable = tmpl.steps.filter(s => {
                // We don't have cl here; trust the DOM's disabled state instead
                const el = document.querySelector(`[data-step-id="${s.id}"]`);
                if (!el) return false;
                return !el.closest('.obc-step-disabled');
            });
            const completed = applicable.filter(s => {
                const v = (_obcCurrentData[s.id] || {}).value;
                return v && v !== '0' && v !== 'false';
            }).length;
            const total = applicable.length;
            const pct = total ? Math.round(completed / total * 100) : 0;
            fill.style.width = pct + '%';
            label.textContent = `${completed} of ${total} applicable steps complete`;
        }

        async function _obcCreateNew(form) {
            const body = {
                client_name: form.client_name.trim(),
                site_url: form.site_url.trim(),
                is_new_client: form.is_new_client,
                is_redesign: form.is_redesign,
                is_new_site: form.is_new_site,
                custom_host: form.custom_host.trim(),
            };
            if (!body.client_name) return;
            const result = await apiCall('/checklists', 'POST', body);
            if (result.id) openObcDetail(result.id);
        }

        registerRoute('checklists', () => loadChecklists());
